### PR TITLE
Add Events for personal feed

### LINKED ISSUE #377: Personal feed requirements
To implement the personal feed in Alchemy we are going to need a new entity in the subgraph that contains all events that could show up in people's feeds. The events we need to track are listed here: https://docs.google.com/document/d/1_Qqoo5m9MbbuGyRG0HUchIFYPrDKU_LKr3T8mTambE4/edit#heading=h.n5awvhl2olm2

The entity can be called `Event`. We wil need to be able to filter events by the objects that they apply to (i.e. if a user is followign dao `0x124`, we want to be able to query all `Events` regarding that dao)

- `type`: one of list below
- `data!` (probably a JSON object with specific data for different types, like for a stake how much was staked - see list below)
- `proposal?` - id of the proposal that the event applies to
- `user?` - the address of the user that initiated the event (staker, voter, proposal creator..). 
- `dao?` - id of the DAO that the event applies to
- `time!` - when the event happened

Here is a list of event types that we want to see [with between square brackets the "index fields"  that are expected to be populated with this eventype:

- `NewProposal` [dao, user (address of proposer), proposal] - data contains `title` 
- `NewReputationHolder` [dao, user ] - data contains `reputationAmount`
- `ProposalStageChange`  [dao, proposal] - data contains the new `stage`
- `Vote` [proposal, user (=address of voter)] - data contains the `Outcome`  and `reputationAmount` (amount of rep)
- `Stake` [proposal, user (=address of staker)] - data contains `stakeAmount` and `Outcome`
-  `VoteFlip` [proposal, user (=address of voter)] - data contains the new `Outcome`, `votesFor`, `votesAgainst`
- `NewDAO` [dao] - data contina the dao's  `address` and  `name`

### PR DESCRIPTION
**Pull request by ben-kaufman**
Fix #377.

### CODE REVIEW

File: src/domain/schema.graphql [hunk 1]
```diff
@@ -188,3 +198,13 @@ type ContractInfo @entity {
 	version: String!
 	address: Bytes!
 }
+
+type Event @entity {
```
Comment (orenyodfat) [inline src/domain/schema.graphql]: how about adding events : [Event!] @derivedFrom('proposal') to proposal entity ?

File: src/domain/index.ts [hunk 2]
```diff
@@ -273,6 +283,19 @@ export function handleStateChange(event: StateChange): void {
           (event.params._proposalState === 2)) {
           insertGPRewards(event.params._proposalId, event.block.timestamp, event.address, event.params._proposalState);
       }
+      let proposal = Proposal.load(event.params._proposalId.toHex());
+      let eventType = 'ProposalStageChange';
+      let eventEntId = crypto.keccak256(concat(event.params._proposalId, event.block.timestamp as ByteArray));
+
+      let eventEnt = new Event(eventEntId.toHex());
+      eventEnt.type = eventType;
+      eventEnt.data = '{ "stage": " ' + proposal.stage + ' " }';
```
Comment (orenyodfat) [inline src/domain/index.ts]: move that to event.ts

File: src/domain/schema.graphql (hunk 1 above)
Comment (ben-kaufman) [reply src/domain/schema.graphql]: @jellegerbrandy will it be useful to add that?

File: src/domain/event.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
+import { Address, BigInt, ByteArray, Bytes, crypto } from '@graphprotocol/graph-ts';
+import { Event, Proposal, ReputationHolder } from '../types/schema';
+import { concat } from '../utils';
+
+export function addNewDAOEvent(avatar: Address, daoName: string, timestamp: BigInt): void {
```
Comment (orenyodfat) [inline src/domain/event.ts]: how about add a function addEvent(address, name,id...) .
- so no need specific function and duplicate code for each event

Comment (ben-kaufman) [reply src/domain/event.ts]: done

File: src/domain/schema.graphql [hunk 4]
```diff
@@ -188,3 +198,13 @@ type ContractInfo @entity {
...
+	id: ID!
+	type: EventType!
+	data: String!
```
Comment (orenyodfat) [inline src/domain/schema.graphql]: what is this data field ? why it is needed ?

Comment (ben-kaufman) [reply src/domain/schema.graphql]: That include some JSON data Alchemy might need. @jellegerbrandy

Comment (orenyodfat) [reply src/domain/schema.graphql]: @jellegerbrandy ?

File: src/domain/event.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    addEvent(
+        'NewDAO',
+        avatar.toHex(),
+        '{ "address": "' + avatar.toHex() + '", "name": "' + daoName + '" }',
```
Comment (orenyodfat) [inline src/domain/event.ts]: what is that ?
why it is hard coded here?

Comment (ben-kaufman) [reply src/domain/event.ts]: That’s just the data object Alchemy needs for this event type...

Comment (orenyodfat) [reply src/domain/event.ts]: on first glance it looks not so elegant solution. is that can be avoided ? @jellegerbrandy

**Comment by orenyodfat** [comment]
Well, is that cannot be analyzed from the event it self ?

File: src/domain/schema.graphql (hunk 4 above)
Comment (orenyodfat) [reply src/domain/schema.graphql]: is that cannot be analyzed from the event it self ?

Comment (orenyodfat) [reply src/domain/schema.graphql]: @jellegerbrandy

File: src/domain/event.ts (hunk 5 above)
Comment (jellegerbrandy) [reply src/domain/event.ts]: tibet is right: the avatar is not strictly needed (although neither does it hurt to have it :-))

I do not really understand @orenyodfat's objection. 
Events have a `data` attribute that is a string. This string happends to be a JSON structure, which is useful when we want to reprseent the event. This does not seem particularly unelegant to me as a pattern

**Comment by orenyodfat** [comment]
Does this data need to be searchable and indexed?
It looks too specific and duplicate the field we  have on the entity it
self .

File: src/domain/event.ts (hunk 5 above)
Comment (orenyodfat) [reply src/domain/event.ts]: Does this data need to be searchable and indexed?
It looks too specific and duplicate the field we  have on the entity it self .

Comment (jellegerbrandy) [reply src/domain/event.ts]: no it does not, that is why it is in a text field and not in its own separate field.

**Pull request merged**